posthog/warehouse/models/managed_viewset.py

```diff
@@ -0,0 +1,210 @@
+                    logger.warning(
+                        "failed_to_delete_orphaned_view",
+                        team_id=self.team_id,
+                        view_name=orphaned_view.name,
+                        error=str(e),
+                    )
+
+            logger.info(
+                "sync_views_completed",
+                team_id=self.team_id,
+                kind=self.kind,
+                views_created=views_created,
+                views_updated=views_updated,
+                views_scheduled=views_scheduled,
+                views_deleted=views_deleted,
+            )
+
+    def delete_with_views(self) -> int:
+        """
+        Delete the managed viewset and soft delete all related views.
+        Reverts materialization for each view before soft deletion.
+        Returns the number of views deleted.
+        """
+        from posthog.warehouse.models.datawarehouse_saved_query import DataWarehouseSavedQuery
+
+        related_views = DataWarehouseSavedQuery.objects.filter(
+            team=self.team,
+            managed_viewset=self,
+        ).exclude(deleted=True)
+
+        views_deleted = 0
+        for view in related_views:
+            try:
+                view.revert_materialization()
+                view.soft_delete()
+                views_deleted += 1
+            except Exception as e:
+                logger.warning(
+                    "failed_to_delete_managed_view",
+                    team_id=self.team_id,
+                    view_name=view.name,
+                    error=str(e),
+                )
+
+        logger.info(
+            "managed_viewset_deleted_with_views",
+            team_id=self.team_id,
+            kind=self.kind,
+            views_deleted=views_deleted,
+        )
+
+        self.delete()
+        return views_deleted
+
+    @staticmethod
+    def _get_clickhouse_type(field) -> str:
+        """Convert HogQL field type to ClickHouse type string."""
+        from posthog.hogql.database.models import (
+            BooleanDatabaseField,
+            DateDatabaseField,
+            DateTimeDatabaseField,
+            FloatDatabaseField,
+            IntegerDatabaseField,
+            StringDatabaseField,
+        )
+
+        if isinstance(field, StringDatabaseField):
+            type = "String"
+        elif isinstance(field, IntegerDatabaseField):
+            type = "Int64"
+        elif isinstance(field, FloatDatabaseField):
+            type = "Float64"
+        elif isinstance(field, DecimalDatabaseField):
+            type = "Decimal64(10)"
+        elif isinstance(field, DateTimeDatabaseField):
+            type = "DateTime64(6, 'UTC')"
+        elif isinstance(field, BooleanDatabaseField):
+            type = "Boolean"
+        elif isinstance(field, DateDatabaseField):
+            type = "Date"
+        else:
+            type = "String"
+
+        if field.is_nullable():
+            type = f"Nullable({type})"
+
+        return type
```

Comment:
These types almost deffo won't map accurately; there are always some nuances with ClickHouse types and we can't really translate hogql types => clickhouse types. Check out `posthog/warehouse/models/table.py#get_columns` - we run a `DESCRIBE TABLE` query on `chdb` to get columns (we also fallback to the prod clickhouse cluster if we need) - but doing something like that will get you far more accurate types

That being said, we very rarely use these values - we check whether the clickhouse type is a `bool` in a couple of places, so this may not matter as much. The types that _really_ matter are those on `DataWarehouseTable` which you're not dealing with, so i think this may be okay. Will leave it up to you to decide

Reply:
Yeah, this is a big code smell here and I don't have a good solution. As you know, it's basically impossible to know what the CH types will actually look like. I tried removing `clickhouse_type` completely, but it is used in a couple of places, so it felt like this was my only option.

I'll add a comment explaining how this is not a good idea, but it's only rarely used anyway. Also, if we ever find an error in here, I made it very easy to resync the `columns` - just call `sync_views` on every managed viewset!